src/value.rs

```diff
@@ -0,0 +1,1613 @@
+            }),
+            Value::I64(v1) => Value::I64(if v2 >= 64 {
+                if v1 < 0 {
+                    !0
+                } else {
+                    0
+                }
+            } else {
+                v1 >> v2
+            }),
+            // It's unclear if unsigned values should implicity convert to a signed value.
+            // For now, we don't support them.
+            Value::U8(_) | Value::U16(_) | Value::U32(_) | Value::U64(_) => {
+                return Err(Error::UnsupportedTypeOperation)
+            }
+            _ => return Err(Error::IntegralTypeRequired),
+        };
+        Ok(value)
+    }
+
+    /// Perform the `==` relational operation.
+    ///
+    /// This operation requires matching integral types.
+    /// If the value type is `Generic`, then it is interpreted as a signed value.
+    ///
+    /// This corresponds to the DWARF `DW_OP_eq` operation.
+    pub fn eq(self, rhs: Value, addr_mask: u64) -> Result<Value> {
+        let value = match (self, rhs) {
+            (Value::Generic(v1), Value::Generic(v2)) => {
+                sign_extend(v1, addr_mask) == sign_extend(v2, addr_mask)
+            }
+            (Value::I8(v1), Value::I8(v2)) => v1 == v2,
+            (Value::U8(v1), Value::U8(v2)) => v1 == v2,
+            (Value::I16(v1), Value::I16(v2)) => v1 == v2,
+            (Value::U16(v1), Value::U16(v2)) => v1 == v2,
+            (Value::I32(v1), Value::I32(v2)) => v1 == v2,
+            (Value::U32(v1), Value::U32(v2)) => v1 == v2,
+            (Value::I64(v1), Value::I64(v2)) => v1 == v2,
+            (Value::U64(v1), Value::U64(v2)) => v1 == v2,
+            (Value::F32(v1), Value::F32(v2)) => v1 == v2,
+            (Value::F64(v1), Value::F64(v2)) => v1 == v2,
+            _ => return Err(Error::TypeMismatch),
+        };
+        Ok(Value::Generic(value as u64))
+    }
+
+    /// Perform the `>=` relational operation.
+    ///
+    /// This operation requires matching integral types.
+    /// If the value type is `Generic`, then it is interpreted as a signed value.
+    ///
+    /// This corresponds to the DWARF `DW_OP_ge` operation.
+    pub fn ge(self, rhs: Value, addr_mask: u64) -> Result<Value> {
+        let value = match (self, rhs) {
+            (Value::Generic(v1), Value::Generic(v2)) => {
+                sign_extend(v1, addr_mask) >= sign_extend(v2, addr_mask)
+            }
+            (Value::I8(v1), Value::I8(v2)) => v1 >= v2,
+            (Value::U8(v1), Value::U8(v2)) => v1 >= v2,
+            (Value::I16(v1), Value::I16(v2)) => v1 >= v2,
+            (Value::U16(v1), Value::U16(v2)) => v1 >= v2,
+            (Value::I32(v1), Value::I32(v2)) => v1 >= v2,
+            (Value::U32(v1), Value::U32(v2)) => v1 >= v2,
+            (Value::I64(v1), Value::I64(v2)) => v1 >= v2,
+            (Value::U64(v1), Value::U64(v2)) => v1 >= v2,
+            (Value::F32(v1), Value::F32(v2)) => v1 >= v2,
+            (Value::F64(v1), Value::F64(v2)) => v1 >= v2,
```

Comment:
Thinking about it a little more, something like this could be useful:

```rust
enum SameTypeValues {
    Generic(u64, u64),
    I8(i8, i8),
    U8(u8, u8),
    I16(i16, i16),
    U16(u16, u16),
    I32(i32, i32),
    U32(u32, u32),
    I64(i64, i64),
    U64(u64, u64),
    F32(f32, f32),
    F64(f64, f64),
}

impl SameTypeValues {
    fn try_from(a: Value, b: Value) -> Result<SameTypeValues> {
        match (a, b) {
            (Generic(x), Generic(y)) => Ok(SameTypeValues::Generic(x, y)),
            (I8(x), I8(y)) => Ok(SameTypeValues::I8(x, y)),
            (U8(x), U8(y)) => Ok(SameTypeValues::U8(x, y)),
            (I16(x), I16(y)) => Ok(SameTypeValues::I16(x, y)),
            (U16(x), U16(y)) => Ok(SameTypeValues::U16(x, y)),
            (I32(x), I32(y)) => Ok(SameTypeValues::I32(x, y)),
            (U32(x), U32(y)) => Ok(SameTypeValues::U32(x, y)),
            (I64(x), I64(y)) => Ok(SameTypeValues::I64(x, y)),
            (U64(x), U64(y)) => Ok(SameTypeValues::U64(x, y)),
            (F32(x), F32(y)) => Ok(SameTypeValues::F32(x, y)),
            (F64(x), F64(y)) => Ok(SameTypeValues::F64(x, y)),
            _ => Err(Error::UnsupportedTypeOperation),
        }
    }
}
```

Reply:
The users of `SameTypeValues` would still have to match on just as many cases though?